src_c/display.c

```diff
@@ -2001,13 +2014,59 @@ pg_toggle_fullscreen(PyObject *self, PyObject *args)
 {
     SDL_Window *win = pg_GetDefaultWindow();
     int result;
+    int ww, wh;
+    SDL_DisplayMode dm;
+
     VIDEO_INIT_CHECK();
     if (!win)
         return RAISE(pgExc_SDLError, "No open window");
-    if (SDL_GetWindowFlags(win) & SDL_WINDOW_FULLSCREEN)
-        result = SDL_SetWindowFullscreen(win, SDL_WINDOW_FULLSCREEN);
-    else
-        result = SDL_SetWindowFullscreen(win, 0);
+
+    SDL_GetWindowSize(win, &ww, &wh);
+    if (SDL_GetDesktopDisplayMode(
+            SDL_GetWindowDisplayIndex(win), &dm) != 0) {
+        return RAISE(pgExc_SDLError, SDL_GetError());
+    }
+
+    if (SDL_GetWindowFlags(win) & SDL_WINDOW_FULLSCREEN_DESKTOP){
+        if (pg_renderer != NULL) {
+            int w, h;
+            SDL_Rect rect;
+            int scale = 1;
+            int xscale, yscale;
+
+            SDL_RenderGetLogicalSize(pg_renderer, &w, &h);
+            xscale = ww / w;
+            yscale = wh / h;
+            scale = xscale < yscale ? xscale : yscale;
+            if (scale < 1)
+                scale = 1;
+
+            result = SDL_SetWindowFullscreen(win, 0);
+            SDL_SetWindowSize(win, w * scale, h * scale);
+            SDL_RenderSetLogicalSize(pg_renderer, w, h);
+        } else {
+            PyObject *surface = pg_GetDefaultWindowSurface();
+            result = SDL_SetWindowFullscreen(win, 0);
+            pgSurface_AsSurface(surface) = SDL_GetWindowSurface(win);
+        }
+    } else {
+        if (pg_renderer != NULL) {
+            int w, h;
+            SDL_Rect rect;
+            SDL_RenderGetLogicalSize(pg_renderer, &w, &h);
+            result = SDL_SetWindowFullscreen(win, SDL_WINDOW_FULLSCREEN_DESKTOP);
```

Comment:
Here it uses `SDL_WINDOW_FULLSCREEN_DESKTOP` in the branch where `SDL_GetWindowFlags(win) & SDL_WINDOW_FULLSCREEN_DESKTOP` is false. Is this intended? I don't understand it well enough to know.

Reply:
Yes. This is the behaviour of SCALED | FULLSCREEN. And in this branch, we have an SDL_Renderer, but the window is **not yet** in FULLSCREEN, so we toggle and set to FULLSCREEN.